src/tracing.rs

```diff
@@ -0,0 +1,364 @@
+
+    fn transaction_number(&self) -> u16 {
+        self.state.transaction_number
+    }
+
+    fn set_transaction_number(&mut self, value: u16) {
+        self.state.transaction_number = value;
+    }
+
+    fn context_u128_register(&self) -> u128 {
+        self.state.context_u128
+    }
+
+    fn set_context_u128_register(&mut self, value: u128) {
+        self.state.context_u128 = value;
+    }
+
+    fn get_storage_state(&self) -> impl Iterator<Item = ((u256::H160, u256::U256), u256::U256)> {
+        self.world_diff
+            .get_storage_state()
+            .iter()
+            .map(|(key, value)| (*key, *value))
+    }
+
+    fn get_storage(&self, _address: u256::H160, _slot: u256::U256) -> Option<(u256::U256, u32)> {
+        todo!() // Do we really want to expose the pubdata?
+    }
+
+    fn get_storage_initial_value(&self, _address: u256::H160, _slot: u256::U256) -> u256::U256 {
+        todo!() // Do we really want to expose the caching?
+    }
+
+    fn write_storage(&mut self, _address: u256::H160, _slot: u256::U256, _value: u256::U256) {
+        todo!()
+    }
+
+    fn get_transient_storage_state(
+        &self,
+    ) -> impl Iterator<Item = ((u256::H160, u256::U256), u256::U256)> {
+        self.world_diff
+            .get_transient_storage_state()
+            .iter()
+            .map(|(key, value)| (*key, *value))
+    }
+
+    fn get_transient_storage(&self, address: u256::H160, slot: u256::U256) -> u256::U256 {
+        self.world_diff
+            .get_transient_storage_state()
+            .get(&(address, slot))
+            .copied()
+            .unwrap_or_default()
+    }
+
+    fn write_transient_storage(
+        &mut self,
+        _address: u256::H160,
+        _slot: u256::U256,
+        _value: u256::U256,
+    ) {
+        todo!()
+    }
+
+    fn events(&self) -> impl Iterator<Item = Event> {
+        self.world_diff.events().iter().map(|event| Event {
+            key: event.key,
+            value: event.value,
+            is_first: event.is_first,
+            shard_id: event.shard_id,
+            tx_number: event.tx_number,
+        })
+    }
+
+    fn l2_to_l1_logs(&self) -> impl Iterator<Item = L2ToL1Log> {
+        self.world_diff.l2_to_l1_logs().iter().map(|log| L2ToL1Log {
+            address: log.address,
+            key: log.key,
+            value: log.value,
+            is_service: log.is_service,
+            shard_id: log.shard_id,
+            tx_number: log.tx_number,
+        })
+    }
+
+    fn pubdata(&self) -> i32 {
+        self.world_diff.pubdata()
+    }
+
+    fn set_pubdata(&mut self, value: i32) {
+        self.world_diff.pubdata.0 = value;
+    }
+
+    fn run_arbitrary_code(_code: &[u64]) {
+        todo!()
+    }
+
+    fn static_heap(&self) -> HeapId {
+        todo!()
+    }
+}
+
+struct CallframeWrapper<'a> {
+    frame: &'a mut Callframe,
+    near_call: Option<usize>,
+}
+
+impl CallframeInterface for CallframeWrapper<'_> {
+    fn address(&self) -> u256::H160 {
```

Comment:
Bikeshedding: At this point it probably doesn't make sense to rename `primitive_types` in `Cargo.toml` because `U256` is not the only type imported.